pkg/endpoint/endpoint.go

```diff
@@ -1458,8 +1458,26 @@ func (e *Endpoint) OnProxyPolicyUpdate(revision uint64) {
 	e.unlock()
 }
 
-// OnDNSPolicyUpdateLocked is called when the Endpoint's DNS policy has been updated
-func (e *Endpoint) OnDNSPolicyUpdateLocked(rules restore.DNSRules) {
+// UpdateDNSRulesIfNeeded is called to synchronize the Endpoint's DNS rules
+// with what stored inside DNS proxy for this endpoint.
+func (e *Endpoint) UpdateDNSRulesIfNeeded() {
+	e.unconditionalLock()
+	// Update DNSRules if any. This is needed because DNSRules also encode allowed destination IPs
+	// and those can change anytime we have identity updates in the cluster. If there are no
+	// DNSRules (== nil) we don't need to update here, as in that case there are no allowed
+	// destinations either.
+	if e.DNSRules != nil {
+		e.OnDNSPolicyUpdateForce(e.owner.GetDNSRules(e.ID))
+		e.getLogger().WithFields(logrus.Fields{
+			"DNSRules": e.DNSRules,
+		}).Debug("Endpoint DNSRules updated. They will be written upon next header file sync")
+	}
+	e.unlock()
+}
+
+// OnDNSPolicyUpdateForce sets the Endpoint's DNS rules forcefully and assumes
+// that the Endpoint write-lock has been taken.
+func (e *Endpoint) OnDNSPolicyUpdateForce(rules restore.DNSRules) {
```

Comment:
nit: the `Force` / `forcefully` wording here seems to be implying something, but I'm not quite sure what. The implementation itself looks like a simple `SetDNSRules(...)` or `UpdateDNSRules(...)`. I think maybe the 'forceful' part is trying to imply that it's the raw setter for the same functionality as `UpdateDNSRulesIfNeeded()`? In which case maybe having more similar names between the two would help to teach the reader that these two functions are variations of the same functionality.

(Evidently the current naming of `OnDNSPolicyUpdateLocked()` was trying to imply that it does more than just setting the rules, but in practice it doesn't. It also pulled in the `Locked` suffix to imply that users must be careful about which locks they hold while calling it. That said I'm not super attached to the existing name given that it seems to be needlessly abstracting away the details just to make the code less specific).

Reply:
Yeah I debated back and forth on this change. What I wanted to do was clarify the implying of 
> it does more than just setting the rules, but in practice it doesn't.

But maybe that's just not worth it. I will revert this change as it's clearly not helpful.

Reply:
Yeah, so-so. Also worth noting that name changes can impact backports, so sometimes it's easier to just leave it alone.